### PR TITLE
net/mlx5_vdpa: Added support for TIS creation using Devx

### PR DESCRIPTION
**Pull request by adrianchiris**
Signed-off-by: Liel Shoshan <liels@mellanox.com>

### CODE REVIEW

File: drivers/net/mlx5/mlx5_vdpa.c [hunk 1]
```diff
@@ -477,6 +483,47 @@ static int mlx5_vdpa_release_rx_steer(struct vdpa_priv *priv)
...
 
+static
+int mlx5_vdpa_create_tis(struct vdpa_priv *priv)
+{
+	uint32_t in[MLX5_ST_SZ_DW(create_tis_in)] = {0};
+	uint32_t out[MLX5_ST_SZ_DW(create_tis_out)] = {0};
+	struct mlx5_vdpa_tx_steer_info *tis = NULL;
+
+	tis = rte_zmalloc("tis", sizeof(*tis), RTE_CACHE_LINE_SIZE);
+	MLX5_SET(create_tis_in, in, opcode, MLX5_CMD_OP_CREATE_TIS);
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_vdpa.c:494]: was the TIS creation successful? which transport domain you used?

File: drivers/net/mlx5/mlx5_vdpa.c [hunk 2]
```diff
@@ -53,23 +53,28 @@ struct mlx5_vdpa_caps {
 };
 
 struct virtq_info {
-	uint32_t                rqn;
-	void                    *rq_buf;
+	uint32_t                qn;
+	void                    *q_buf;
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_vdpa.c:57]: we are going to remove the rq soon, but why this change?

**Review by shahafsh — CHANGES_REQUESTED** [review]
Hi Liel, 

see some comments.

File: drivers/net/mlx5/mlx5_prm.h [hunk 3]
```diff
@@ -1275,4 +1276,37 @@ struct mlx5_ifc_create_mkey_in_bits {
...
+	u8         reserved_at_8[0x18];
+	u8         syndrome[0x20];
+	u8         reserved_at_40[0x8];
+	u8         tisn[0x18];
+	u8         reserved_at_60[0x20];
+};
+
+struct mlx5_ifc_create_tis_in_bits {
+	u8         opcode[0x10];
+	u8         uid[0x10];
```
Comment (shamoya) [inline drivers/net/mlx5/mlx5_prm.h:1305]: Like we did in all other commands Input structures, let's put reserved instead of uid (as it's not exposed to user)

File: drivers/net/mlx5/mlx5_vdpa.c (hunk 2 above)
Comment (shamoya) [reply drivers/net/mlx5/mlx5_vdpa.c:57]: @shahafsh , this is what we discussed before the change was made

File: drivers/net/mlx5/mlx5_vdpa.c [hunk 4]
```diff
@@ -389,11 +395,11 @@ create_split_virtq(struct vdpa_priv *priv, int index,
 	if (is_virtq_recvq(index, priv->nr_vring)) {
 		MLX5_SET(virtq, virtq, queue_type,
 			 MLX5_VIRTQ_OBJ_QUEUE_TYPE_RX);
-		MLX5_SET(virtq, virtq, qnum, priv->virtq[index].rqn);
+		MLX5_SET(virtq, virtq, qnum, priv->virtq[index].qn);
 	} else {
 		MLX5_SET(virtq, virtq, queue_type,
 			  MLX5_VIRTQ_OBJ_QUEUE_TYPE_TX);
-		/* TODO(liel): Please assign here the SQN to qnum */
+		MLX5_SET(virtq, virtq, qnum, priv->virtq[index].qn);
```
Comment (shamoya) [inline drivers/net/mlx5/mlx5_vdpa.c:402]: Although this code is removed in new PRM, it was nicer to have it out of the if/else as it's the same line

File: drivers/net/mlx5/mlx5_vdpa.c [hunk 5]
```diff
@@ -525,14 +576,18 @@ static int mlx5_vdpa_release_virtqs(struct vdpa_priv *priv)
 		DRV_LOG(ERR, "Error Releasing RX steering resources");
 		return -1;
 	}
+	if (mlx5_vdpa_release_tx_steer(priv)) {
+		DRV_LOG(ERR, "Error Releasing TX steering resources");
+		return -1;
+	}
```
Comment (shamoya) [inline drivers/net/mlx5/mlx5_vdpa.c:582]: Releasing should be done in LIFO, so this should be on the bottom

File: drivers/net/mlx5/mlx5_vdpa.c [hunk 6]
```diff
@@ -477,6 +483,47 @@ static int mlx5_vdpa_release_rx_steer(struct vdpa_priv *priv)
...
 }
 
+static
+int mlx5_vdpa_create_tis(struct vdpa_priv *priv)
+{
+	uint32_t in[MLX5_ST_SZ_DW(create_tis_in)] = {0};
+	uint32_t out[MLX5_ST_SZ_DW(create_tis_out)] = {0};
+	struct mlx5_vdpa_tx_steer_info *tis = NULL;
+
+	tis = rte_zmalloc("tis", sizeof(*tis), RTE_CACHE_LINE_SIZE);
```
Comment (shamoya) [inline drivers/net/mlx5/mlx5_vdpa.c:493]: I don't understand why do we need a malloced mlx5_vdpa_tx_steer_info here
You can just do the create_tis and put the devx_obj and TISN in priv->tx_steer_info right?

File: drivers/net/mlx5/mlx5_vdpa.c [hunk 7]
```diff
@@ -477,6 +483,47 @@ static int mlx5_vdpa_release_rx_steer(struct vdpa_priv *priv)
...
+	MLX5_SET(create_tis_in, in, opcode, MLX5_CMD_OP_CREATE_TIS);
+	tis->tis_obj =
+		mlx5_glue->dv_devx_obj_create(priv->ctx, in,
+					sizeof(in), out, sizeof(out));
+	if (!tis) {
```
Comment (shamoya) [inline drivers/net/mlx5/mlx5_vdpa.c:498]: Should be tis->tis_obj in the check

**Comment by shamoya** [comment]
Thanks @bekets 
I think after recent PRM changes, no point in merging this for now.
Maybe you can have the create_tis done with the rest of the changes

File: drivers/net/mlx5/mlx5_vdpa.c (hunk 1 above)
Comment (bekets) [reply drivers/net/mlx5/mlx5_vdpa.c:494]: It was, without any transport domain there is just no enforcement on LB traffic

File: drivers/net/mlx5/mlx5_vdpa.c (hunk 4 above)
Comment (bekets) [reply drivers/net/mlx5/mlx5_vdpa.c:402]: right, moved it out of if/else

File: drivers/net/mlx5/mlx5_prm.h (hunk 3 above)
Comment (bekets) [reply drivers/net/mlx5/mlx5_prm.h:1305]: changed it to reserved_at_10

File: drivers/net/mlx5/mlx5_vdpa.c (hunk 6 above)
Comment (bekets) [reply drivers/net/mlx5/mlx5_vdpa.c:493]: right, changed

File: drivers/net/mlx5/mlx5_vdpa.c (hunk 7 above)
Comment (bekets) [reply drivers/net/mlx5/mlx5_vdpa.c:498]: right ,thanks

**Comment by shamoya** [comment]
Already in
